### DISCUSSION TITLE
How to make android emulator start with the right width?

### DISCUSSION THREAD

**Discussion by ivopr** [qtile/qtile#4454] [Q&A] [answered]
So, whenever I open android emulator It gets like this

![image](https://github.com/qtile/qtile/assets/30270448/e0dfa736-daa1-489b-a1c2-443d6226fbd4)

And then I do [cmd]+Shift+Right so I can resize it to this

![image](https://github.com/qtile/qtile/assets/30270448/ecdbd935-c8e3-4d1a-831a-5c28f728d059)

Is there any way that I can say to Qtile to aways start it with the last image width? or any fixed width that I define?

**Comment by ervinpopescu** [comment]
What layout are you using?

**Reply by ivopr** [reply]
forgot about that, it's the Columns layout

**Reply by ervinpopescu** [reply]
What is the layout configuration?

**Reply by ivopr** [reply]
@ervinpopescu you mean thos block here?
```py
# L A Y O U T S
layouts = [
    layout.Columns(
        margin=3,
        border_focus='#607767',
	    border_normal='#1F1D2E',
        border_width=2,
    )
]
```

**Reply by ervinpopescu** [reply]
I don't think there is any way other than adding a hook that checks for one of the window attributes (title, class) and runs window.layout.grow_right() on the terminal window that opened the android emulator window. Will try to write something.

**Reply by ivopr** [reply]
yeah, I got halfway on the Hook, Just failed to set the size. do we have any shrink_left option on the window? I think it'll be easier to make the emulator shrink when It opens

**Reply by ervinpopescu** [reply]
Unfortunately not, you probably need to hack a solution to this by using `layout.previous()` -> `layout.grow_right()` -> `layout.next()`.

**Reply by ivopr** [reply]
I've managed to get to

```py
    if window_wm_name is "Android Emulator - Pixel_3a_API_34_extension_level_7_x86_64:5554":
        # layout = qtile.current_layout
        layout = window.layout
        layout.previous()
        layout.grow_right()
        layout.grow_right()
        layout.grow_right()
        layout.grow_right()
```

but I'm not even sure if these things exist. And no, it doesn't work

**Reply by ervinpopescu** [reply]
How about `window.group.current_layout`?

**Reply by ivopr** [reply]
didn't work. man, that's really difficult

**Reply by ervinpopescu** [reply]
Yeah, navigating qtile's graph takes a better mind than mine. @elParaguayo, can you please help when you're done carrying some of Qtile's Atlassian globe on your shoulders? /s

**Reply by ervinpopescu** [reply]
Okay, apologies. I always mistake `current_layout` with `layout`. `current_layout` gives the index of the current layout in the `layouts` attribute of the `group` and `layout` actually returns the object (see [this](https://github.com/qtile/qtile/blob/master/libqtile/group.py#L112-L113)). Please retry using `window.group.layout`.

**Reply by ivopr** [reply]
yep, didn't work too. Also, that's me reminiscing from some time ago, maybe it is the title of the android studio emulator. I'm not sure it is set correctly at the time it spawns. I vaguely remember it being empty or 'Android Emulator' for a bit before setting to 'Android Emulator - emulator name here'

**Reply by ivopr** [reply]
Like, how could I `print` something in a hook and see it? for me at least that's the thing blocking me from going further.

There's qtile's send_notification too, thought about using it, but, where does this notification go?

**Reply by ervinpopescu** [reply]
Well, you can use `logger.debug()` from `libqtile.log_utils`, or set the logging level to `INFO`, by using `qtile.info()` and then use `logger.info("%s", whatever_you_want_to_log)`. And the notification is sent to a notification server, like `dunst`, which you need to have running in order to see the notification.

**Reply by ervinpopescu** [reply]
you could also try to get the class, instead of the name, with `client.window.get_wm_class()[0]`, where `client` is the argument of the `hook`ed function.

**Reply by ivopr** [reply]
the android emulator doesn't have a wm class assigned to it. that makes me kinda sad. some time ago I was all about making it float and whatnot and discovered that by chance. Interestingly, it did make it float finding it by name. I'll try logging things here.

**Reply by ervinpopescu** [reply]
ah okay, maybe you can `sleep` for a while and then check out the window name?

**Reply by ivopr** [reply]
yeah, I'll try things out here now that I'm able to see the logs, will update here soon

**Reply by ivopr** [reply]
oh, ok, so, I've checked here, and I'm on 0.22.1, would this impede something we discussed previously here? using PyCharm now to get the right intellisense as VSCode just refused to, I get that client.group.layout doesn't have `previous()` or any of the `grow_` methods

**Reply by elParaguayo** [reply]
I've only got a really hacky suggestion (and I haven't read the whole chain here): you can use a function to specify the column widths. You may need to call this with a key binding if you can't use a hook.

The other problem is, you can't set the column width for a column that doesn't exist yet so you'd need to add the second column first then you could do something like:

```python
@lazy.layout.function
def resize_columns(layout):
    if layout.name != "columns":
        return

    layout.columns[0] = 75 
    layout.columns[1] = 25
    layout.group.layout_all()
```
Then you can add `resize_columns()` to a key binding.

**Reply by ervinpopescu** [reply]
> oh, ok, so, I've checked here, and I'm on 0.22.1, would this impede something we discussed previously here? using PyCharm now to get the right intellisense as VSCode just refused to, I get that client.group.layout doesn't have `previous()` or any of the `grow_` methods

Yeaaah, this is a PITA, you need to `type` everything correctly, like `client: Window` etc.

@elParaguayo this is a neat workaround, but I still think this can be done without needing to bind it to a key and triggering it manually...

**Reply by ervinpopescu** [reply]
> oh, ok, so, I've checked here, and I'm on 0.22.1, would this impede something we discussed previously here? using PyCharm now to get the right intellisense as VSCode just refused to, I get that client.group.layout doesn't have `previous()` or any of the `grow_` methods

oh, I really am not familiar with that version of qtile anymore. there still is [`cmd_previous`](https://github.com/qtile/qtile/blob/v0.22.1/libqtile/layout/columns.py#L387) and [`cmd_grow_`](https://github.com/qtile/qtile/blob/v0.22.1/libqtile/layout/columns.py#L460).

**Reply by elParaguayo** [reply]
> @elParaguayo this is a neat workaround, but I still think this can be done without needing to bind it to a key and triggering it manually...

Agreed - but the normal way would be to use one of the hooks (e.g. `client_managed`) for this. However, if you can't match the window title or window class it becomes a bit harder to do.

**Reply by ivopr** [reply]
Just checked and it seems the window title is correctly set on startup, tried forcing it to open in a group using this

```py
dgroups_app_rules = [
    Rule(Match(wm_class="code"), group="1"),
    Rule(Match(wm_class="Alacritty"), group="2"),
    Rule(Match(wm_class="google-chrome"), group="3"),
    Rule(Match(wm_title="Android Emulator long title her"), group="4"),
]
```

Not sure if it validates that the window title is set from the start but I think (and hope) so. Also, we can use hooks, I've been trying with `client_new`

**Reply by ivopr** [reply]
ok, after some ancient level debugging (as I wasn't able to use logger.debug or logger.info due to my own problems), I've found out that Android Emulator truly has another title at creation: "Emulator", so the first part is ok and we arrived here:

```py
@hook.subscribe.client_new
def on_new_client(client: Window):
    floating_window_names = ["Save File", "Open File", "Welcome to Android Studio"]

    window_wm_name = client.name
    with open("/home/ivopr/Git/dotfiles/log.txt", "w") as f:
        f.write(f"NO IF WINDOW WM NAME: {window_wm_name}")
    if window_wm_name == "Emulator":
        with open("/home/ivopr/Git/dotfiles/log.txt", "w") as f:
            f.write(f"WINDOW WM NAME: {window_wm_name}")
>>>>>>>>>>>>> we're here
        client.group.layout.cmd_previous()
        client.group.layout.cmd_grow_right()
        client.group.layout.cmd_grow_right()
        client.group.layout.cmd_grow_right()
        client.group.layout.cmd_grow_right()
```

the commands below my incredible debugging methodology either work but I need to wait them (?) or don't as the result is ultimately the same

ps.: no one saw that `floating_window_names` there

**Reply by elParaguayo** [reply]
You could try the `client_managed` hook instead as that gets fired slightly later than `client_new`. Also, instead of your commands you could use the lines in my earlier code i.e.
```python
layout = client.group.layout
layout.columns[0] = 75 
layout.columns[1] = 25
layout.group.layout_all()
```

**Reply by ivopr** [reply, marked as answer]
hey @elParaguayo, @ervinpopescu thank you guys for being patient on this thread and sharing somethings. Took some day off the problem, and miraculously (it's magic, I tell you) today I've figured out how to make it work.

First of all, I've migrated to qtile `0.23.0b3.dev8+g133c4119` (I don't remember anymore but I think I've installed it with `pipx`).

Secondly, it appears that `client.group.layout` is not the layout object (see the image)
![image](https://github.com/qtile/qtile/assets/30270448/73c96b2b-f50f-41a9-9854-b9cbd8584d10)

The first approach, proposed by @ervinpopescu did it as soon as I has the layout object, which I got with `client.group.layouts` and `client.group.current_layout`. At the end of the day, the code's below, and again, thank you for the patience invested in me

```py
@hook.subscribe.client_managed
def resize_android_emulator(client: Window):
    if client.name == "Emulator":
        group_layout = client.group.layouts[client.group.current_layout]
        group_layout.previous()
        group_layout.grow_right()
        group_layout.grow_right()
        group_layout.grow_right()
        group_layout.grow_right()
```
Also, I think it doesn't really matters if this code is on `client_managed` or `client_new` hook since the emulator is still named 'Emulator' at both points.

**Comment by ivopr** [comment]
@elParaguayo with this block

```py
@hook.subscribe.client_managed
def resize_android_emulator(client: Window):
    if client.name == "Emulator":
        client.group.layout.columns[0] = 75
        client.group.layout.columns[1] = 25
        client.group.layout.group.layout_all()
```

this is the result.
![image](https://github.com/qtile/qtile/assets/30270448/a280e5f4-d9b9-4e1d-ace9-3a31744b4a2b)

Also, on both, `client_new` and `client_managed` the client name is 'Emulator', that's not really the problem, actually is a benefit as every single android emulator will be spawn named 'Emulator' so it'll apply to any emulator I create later

edit: yay replied on the wrong box
